### ISSUE TITLE
Please work harder to figure out parameter types for sequences

### ISSUE THREAD

**Issue by jimfulton** [sqlalchemy/sqlalchemy#6874]
### Describe the use case

BigQuery needs to know parameter types, even for `None`/`null`.  

When passing a sequence of values:

```
>>> sqlalchemy.bindparam("foo", value=(1, None), expanding=True)
BindParameter('foo', (1, None), type_=Integer())
>>> sqlalchemy.bindparam("foo", value=(None, 1), expanding=True)
BindParameter('foo', (None, 1), type_=NullType())

```
The parameter type depends on the order of values if there are `None`s present.

This seems unsat. :)

The problem is that the only the fist value is inpected:

https://github.com/sqlalchemy/sqlalchemy/blob/70a33dd32f2609f2de6421c3b28ef4cc4496fdae/lib/sqlalchemy/sql/elements.py#L1530-L1533 

I think the first non-null value (if any) should be used.

I'd be happy to submit a PR.

### Databases / Backends / Drivers targeted

BigQuery

### Example Use

```
>>> sqlalchemy.bindparam("foo", value=(1, None), expanding=True)
BindParameter('foo', (1, None), type_=Integer())
>>> sqlalchemy.bindparam("foo", value=(None, 1), expanding=True)
BindParameter('foo', (None, 1), type_=NullType())

```

**Comment by CaselIT** [comment]
Hi,

I think it would makes sense, but I'm not sure about the proposed approach:

> I think the first non-null value (if any) should be used.

potentially very long sequences may be passed, so I think we should have an upper bound. something check the first 100 for a not null, else bail?

**Comment by zzzeek** [comment]
I need more background on the use case, as there was not an intention for "None" to be present in an IN;   it's not really useful from a SQL point of view to have NULL inside of an IN list.  in most cases, it can't change the result of an expression and in one case it only makes the expression behave more poorly.

testing for NULL in the list always returns NULL:

```
test=> select NULL in (1, 2, 3);
 ?column? 
----------
 
(1 row)

test=> select NULL in (1, 2, 3, NULL);
 ?column? 
----------
 
(1 row)

test=> select NULL not in (1, 2, 3);
 ?column? 
----------
 
(1 row)
```

for a non-null member that's in the list, NULL does not affect the value:

```
test=> select 1 in (1, 2, 3, NULL);
 ?column? 
----------
 t
(1 row)

test=> select 1 in (1, 2, 3);
 ?column? 
----------
 t
(1 row)
```

NULL does affect, negatively, if a non-null value is *not* in the list, which is a reason to not include NULL in an IN list:

```
test=> select 5 in (1, 2, 3);
 ?column? 
----------
 f
(1 row)

test=> select 5 in (1, 2, 3, NULL);
 ?column? 
----------
 
(1 row)
```

so if the use case is, "NULL in an IN list", I just need to see more specifics on that since this is IMO not something anyone would want to do, and if it is, I'd want to add real tests for it.

However, the title of this bug report doesn't mention NULL, that's only the example given.    If OTOH you are trying to support something like `(1, "some string", True)` for some kind of non "IN" case,  that's not the "expanding IN" use case; instead, that sounds like you'd want to use the tuple datatype which does in fact type the individual elements.

so I just need context on what it is we're trying to do here.  to my knowledge we dont have any tests which hit that codepath with a "None" in the list.

**Comment by zzzeek** [comment]
> Hi,
> 
> I think it would makes sense, but I'm not sure about the proposed approach:
> 
> > I think the first non-null value (if any) should be used.
> 
> potentially very long sequences may be passed, so I think we should have an upper bound. something check the first 100 for a not null, else bail?

assuming "NULL in an IN list" is just a bad idea, this would be an expensive approach, we can instead just block NULL from being present and/or raise / warn / bail when we render the parameters.

**Comment by CaselIT** [comment]
> assuming "NULL in an IN list" is just a bad idea,

I guess expanding is only used for the in, right?

**Comment by jimfulton** [comment]
> assuming "NULL in an IN list" is just a bad idea, this would be an expensive approach, we can instead just block NULL from being present and/or raise / warn / bail when we render the parameters.

You don't need to assume anything.  You already grab the first value. Checking if it's `None` is easy enough.

Maybe `None` shouldn't be in sequences given to `in_`, but it seems a stretch to assume it won't happen. :shrug: 

In any case, you don't have to scan the input looking for `None`, I'm just suggesting checking if `check_value` is `None` and trying to find a better one if it is.  If you say that this is so unlikely that you don't want to incur the expense, I respect that.

In the BigQuery case, the consequence of not being able to determine parameter types seems to be much greater than for other databases.  I can probably tackle this in the BigQuery dialect's `visit_bindparam`.

**Comment by zzzeek** [comment]
> > assuming "NULL in an IN list" is just a bad idea, this would be an expensive approach, we can instead just block NULL from being present and/or raise / warn / bail when we render the parameters.
> 
> You don't need to assume anything. You already grab the first value. Checking if it's `None` is easy enough.

right, sorry, I was not grounded in this issue because I wasnt sure if you were coming from a case with heterogeneous types in the list, or just None, and then if somehow bigquery needed to differentiate the ints from nones in the bound parameters which would also be weird, so when @CaselIT  said "we just check the first 100" I was seeing a scan of 100 values for any None.    the suggested approach here is, "continue testing until we get a non-None", which for the whole world will be just one element anyway, so there's no issue there.   it's easy, no problem.

> In the BigQuery case, the consequence of not being able to determine parameter types seems to be much greater than for other databases. I can probably tackle this in the BigQuery dialect's `visit_bindparam`.

no this is def. a bug on our end because if the IN is against some datatype that needs typing support like UUID we can easily show this failing.

**Comment by zzzeek** [comment]
OK well, im trying to create a test case and it's not happening, because when we have an expression (x IN y), if "y" has NullType we infer the type from the left side.  So I can't make the UUID case fail here.   any chance I can see the context how you are seeing this create a problem ?

**Comment by CaselIT** [comment]
> so when @CaselIT said "we just check the first 100" I was seeing a scan of 100 values for any None.

no, I meant

> continue testing until we get a non-None and in any case at most 100 element

where 100 is just a random limit value.

**Comment by zzzeek** [comment]
> > so when @CaselIT said "we just check the first 100" I was seeing a scan of 100 values for any None.
> 
> no, I meant
> 
> > continue testing until we get a non-None and in any case at most 100 element
> 
> where 100 is just a sensible (?) limit value.

yeah i get that.  I think we should just keep scanning to the end as this is not at all a normal occurrence.

just at the moment trying to see how this particular typing decision actually affects things as i cant get a failing test case yet.

**Comment by zzzeek** [comment]
hmm.  OK well first off the auto-coercion thing there only counts for the basic types int/str/date/boolean, so UUID is not affected by this.      so here is the test case:

```py
import datetime

from sqlalchemy import literal


expr = literal("a string").in_([datetime.datetime.today()])

print(expr.left.type)
print(expr.right.type)

print("-------")

expr = literal("a string").in_([None, datetime.datetime.today()])

print(expr.left.type)
print(expr.right.type)
```

output:

```
VARCHAR
DATETIME
-------
VARCHAR
NULL
```
